File: langs/python/src/vmx/forms/form_vm.py
"""FormVM[TM] — snapshot/revert edit lifecycle ViewModel.

See spec/20-form-vm.md and ADR-0030.
"""

from __future__ import annotations

import asyncio
import copy
from collections.abc import Awaitable, Callable, Mapping
from typing import Any, Generic, TypeVar

import reactivex as rx
from reactivex import operators as ops
from reactivex.subject import Subject

from vmx.commands.relay_command import RelayCommand
from vmx.forms.builders import FormVMBuilder
from vmx.messages.form_reverted import FormRevertedMessage
from vmx.messages.property_changed import PropertyChangedMessage
from vmx.services.message_hub import MessageHubProto
from vmx.services.null_message_hub import NULL_MESSAGE_HUB

TM = TypeVar("TM")

FieldValidator = Callable[[TM], str | None]
ModelValidator = Callable[[TM], Mapping[str, str | None]]


class FormVM(Generic[TM]):
    """ViewModel that wraps a mutable domain model with an edit lifecycle.

    Captures a snapshot at construction; allows mutation via :meth:`set_model`;
    provides :attr:`deny_command` (revert) and :attr:`approve_command` (persist).

    Parameters
    ----------
    initial:
        Initial domain model; also becomes the initial :attr:`snapshot`.
    persister:
        Async callable ``(model) -> Awaitable[None]``.  Raise on failure.
    hub:
        Message hub.  Defaults to ``NULL_MESSAGE_HUB`` (no hub messages).
    strict:
        When ``True``, ``approve_command.can_execute()`` returns ``False``
        when ``is_dirty`` is ``False``.  Default ``False``.
    snapshotter:
        Custom snapshot function ``(model) -> model``.  Defaults to
        ``copy.deepcopy`` so that a mutation to a *nested* object inside the
        model is visible to :attr:`is_dirty` and restored by ``deny``/revert.
        Inject a custom snapshotter as the escape hatch for models that
        ``deepcopy`` cannot handle (e.g. live handles, unpicklable objects).
    reset_on_approved:
        Optional synchronous callback that derives the next pristine model
        from the captured value after persistence succeeds.
    """

    def __init__(
        self,
        initial: TM,
        persister: Callable[[TM], Awaitable[None]],
        hub: MessageHubProto[Any] | None = None,
        strict: bool = False,
        snapshotter: Callable[[TM], TM] | None = None,
        validators: Mapping[str, FieldValidator[TM]] | None = None,
        model_validator: ModelValidator[TM] | None = None,
        reset_on_approved: Callable[[TM], TM] | None = None,
    ) -> None:
        if initial is None:
            raise ValueError("initial must not be None")
        if persister is None:
            raise ValueError("persister must not be None")

        self._persister = persister
        self._hub: MessageHubProto[Any] = hub if hub is not None else NULL_MESSAGE_HUB
        self._strict = strict
        self._snapshotter: Callable[[TM], TM] = (
            snapshotter if snapshotter is not None else copy.deepcopy
        )
        self._validators: dict[str, FieldValidator[TM]] = dict(validators or {})
        self._model_validator = model_validator
        self._reset_on_approved = reset_on_approved

        self._model: TM = initial
        self._snapshot: TM = self._snapshotter(initial)
        self._errors: dict[str, str] = self._validate(initial)

        self._disposed = False

        # Observables
        self._on_approved: Subject[TM] = Subject()
        self._approve_errors: Subject[BaseException] = Subject()
        self._errors_changed: Subject[dict[str, str]] = Subject()
        self._can_execute_trigger: Subject[None] = Subject()

        # Commands
        self._deny_command = RelayCommand.builder().task(self._deny).build()
        self._approve_command = (
            RelayCommand.builder()
            .task(self._approve_fire_and_forget)
            .predicate(lambda: self.is_valid and (not self._strict or self.is_dirty))
            .triggers(self._can_execute_trigger)
            .build()
        )

    # ── Builder entrypoint ───────────────────────────────────────────────────

    @staticmethod
    def builder() -> FormVMBuilder[TM]:
        """Return a fresh :class:`FormVMBuilder` for this generic type.

        Provided for parity with the other VMx VM family entry points.
        See ADR-0035 §2 FV1. (Instantiated bare, like every sibling
        ``builder()``: subscripted instantiation of a frozen+slots dataclass
        raises ``TypeError`` when typing assigns ``__orig_class__``.)
        """
        return FormVMBuilder()

    @property
    def model(self) -> TM:
        """Live, editable model."""
        return self._model

    @property
    def snapshot(self) -> TM:
        """Read-only snapshot captured at construction (until next approve)."""
        return self._snapshot

    @property
    def is_dirty(self) -> bool:
        """``True`` when ``model`` is structurally not equal to ``snapshot``."""
        return self._model != self._snapshot

    @property
    def errors(self) -> dict[str, str]:
        """Current validation errors keyed by field name."""
        return dict(self._errors)

    @property
    def is_valid(self) -> bool:
        """``True`` when the current model has no validation errors."""
        return not self._errors

    @property
    def deny_command(self) -> RelayCommand:
        """Reverts ``model`` to ``snapshot`` and publishes hub messages."""
        return self._deny_command

    @property
    def approve_command(self) -> RelayCommand:
        """Invokes the persister; on success advances ``snapshot`` and fires ``on_approved``."""
        return self._approve_command

    @property
    def on_approved(self) -> rx.Observable[TM]:
        """Observable that emits the current model after each successful persist."""
        return self._on_approved.pipe(ops.as_observable())

    @property
    def approve_errors(self) -> rx.Observable[BaseException]:
        """Observable that surfaces the persister error when the approve *command* fails.

        ``approve_command.execute()`` is fire-and-forget, so a persister failure
        cannot propagate to the caller; it is emitted here instead of being
        swallowed (VMX-008). The awaitable :meth:`approve_async` keeps its
        raising behavior — ``await`` it directly to handle the error inline.
        Completes on :meth:`dispose`.
        """
        return self._approve_errors.pipe(ops.as_observable())

    @property
    def errors_changed(self) -> rx.Observable[dict[str, str]]:
        """Observable that emits when the effective validation error map changes."""
        return self._errors_changed.pipe(ops.as_observable())

    def field_error(self, field: str) -> str | None:
        """Return the current validation error for ``field``, if any."""
        return self._errors.get(field)

    # ── Mutation ──────────────────────────────────────────────────────────────

    def set_model(self, model: TM) -> None:
        """Replace the current model.

        If in strict mode and ``is_dirty`` changes, fires ``can_execute_changed``.
        Inert after :meth:`dispose` (like :meth:`approve_async`/:meth:`deny`): a
        post-dispose call would otherwise ``on_next`` a disposed reactivex Subject
        via ``_revalidate`` and raise (parity with the TS/Swift no-op).
        """
        if self._disposed:
            return
        if model is None:
            raise ValueError("model must not be None")
        was_dirty = self.is_dirty
        was_valid = self.is_valid
        self._model = model
        self._revalidate()
        if (self._strict and self.is_dirty != was_dirty) or self.is_valid != was_valid:
            self._can_execute_trigger.on_next(None)

    # ── Async core ────────────────────────────────────────────────────────────

    async def approve_async(self) -> None:
        """Awaitable entry-point to the approve flow.

        Invokes the persister, advances :attr:`snapshot` on success, and fires
        :attr:`on_approved`.  Raises when the persister raises (no state mutation).
        A disposed form is a full no-op — the persister is not invoked
        (symmetric with the deny guard).
        """
        if self._disposed:
            return
        if not self.is_valid:
            return
        current = self._model

        # May raise — intentional.  No state mutation if this raises.
        await self._persister(current)

        # dispose() may have run during the await; the subjects below are
        # completed and disposed, so emitting would raise DisposedException
        # (mirrors the C# guard).
        if self._disposed:
            return

        # Success: atomically install the configured reset state, or preserve
        # the legacy snapshot-advance behavior when no reset is configured.
        was_dirty = self.is_dirty
        was_valid = self.is_valid
        if self._reset_on_approved is not None:
            # Prepare callback output, independent live/snapshot values, and
            # validation before committing any local state.
            reset = self._reset_on_approved(current)
            next_model = self._snapshotter(reset)
            next_snapshot = self._snapshotter(reset)
            next_errors = self._validate(next_model)

            self._model = next_model
            self._snapshot = next_snapshot
            if next_errors != self._errors:
                self._errors = next_errors
                self._errors_changed.on_next(dict(next_errors))
        else:
            self._snapshot = self._snapshotter(current)

        if (self._strict and self.is_dirty != was_dirty) or self.is_valid != was_valid:
            self._can_execute_trigger.on_next(None)

        # Emit the value that was actually persisted (parity with C#'s
        # captured `current`): a set_model racing the persister await must
        # not swap the approved payload for a newer un-persisted model.
        self._on_approved.on_next(current)

    # ── Dispose ───────────────────────────────────────────────────────────────

    def dispose(self) -> None:
        """Complete the ``on_approved`` observable and dispose resources. Idempotent."""
        # reactivex Subjects raise DisposedException on a second on_completed,
        # unlike rxjs (no-op) and the guarded C# FormVM — guard for parity.
        if self._disposed:
            return
        self._disposed = True
        self._on_approved.on_completed()
        self._on_approved.dispose()
        self._approve_errors.on_completed()
        self._approve_errors.dispose()
        self._errors_changed.on_completed()
        self._errors_changed.dispose()
        self._can_execute_trigger.on_completed()
        self._can_execute_trigger.dispose()
        self._deny_command.dispose()
        self._approve_command.dispose()

    # ── Internal ──────────────────────────────────────────────────────────────

    def _deny(self) -> None:
        if self._disposed:
            return
        was_dirty = self.is_dirty
        was_valid = self.is_valid
        self._model = self._snapshotter(self._snapshot)
        self._revalidate()

        self._hub.send(FormRevertedMessage(sender=self, sender_name="FormVM"))
        self._hub.send(
            PropertyChangedMessage.create(
                sender=self,
                sender_name="FormVM",
                property_name="model",
            )
        )

        if (self._strict and was_dirty != self.is_dirty) or self.is_valid != was_valid:
            self._can_execute_trigger.on_next(None)

    def _validate(self, model: TM) -> dict[str, str]:
        errors: dict[str, str] = {}
        for field, validator in self._validators.items():
            error = validator(model)
            if error is not None:
                errors[field] = error
        if self._model_validator is not None:
            for field, error in self._model_validator(model).items():
                if error is None:
                    errors.pop(field, None)
                else:
                    errors[field] = error
        return errors

    def _revalidate(self) -> None:
        errors = self._validate(self._model)
        if errors == self._errors:
            return
        self._errors = errors
        self._errors_changed.on_next(dict(errors))

    def _approve_fire_and_forget(self) -> None:
        """Synchronous wrapper that schedules the async approve.

        Called by the RelayCommand (which expects a sync callable). Fire-and-forget:
        a persister failure is routed to :attr:`approve_errors` instead of being
        swallowed (VMX-008). Use :meth:`approve_async` for awaitable raising behavior.
        """
        try:
            loop = asyncio.get_running_loop()
        except RuntimeError:
            # No running loop: run to completion synchronously, routing a
            # persister failure to the error channel (parity with the
            # running-loop path rather than propagating to the sync caller).
            try:
                asyncio.run(self.approve_async())
            except Exception as exc:
                self._emit_approve_error(exc)
            return
        task = loop.create_task(self.approve_async())
        # Mirror the C# continuation: surface the persister failure on the error
        # channel. Skip cancelled tasks — Task.exception() raises CancelledError
        # for those, which would error the loop's callback.
        task.add_done_callback(self._on_approve_done)

    def _on_approve_done(self, task: asyncio.Future[None]) -> None:
        if task.cancelled():
            return
        exc = task.exception()
        if exc is not None:
            self._emit_approve_error(exc)

    def _emit_approve_error(self, exc: BaseException) -> None:
        # The subject is completed+disposed on dispose(); a persister failure
        # arriving after dispose must not raise reactivex DisposedException.
        if self._disposed:
            return
        self._approve_errors.on_next(exc)

File: langs/csharp/src/VMx/Forms/FormVM.cs
using System.Reactive;
using System.Reactive.Linq;
using System.Reactive.Subjects;
using System.Windows.Input;
using VMx.Commands;
using VMx.Internal;
using VMx.Messages;
using VMx.Services;

namespace VMx.Forms;

/// <summary>
/// ViewModel that wraps a mutable domain model with an edit lifecycle:
/// snapshot on construct, allow mutation, then either Approve (persist) or Deny (revert).
///
/// See spec/20-form-vm.md and ADR-0030.
/// </summary>
/// <typeparam name="TM">Domain model type. Record types are recommended for structural equality.</typeparam>
public sealed class FormVM<TM> : IDisposable
    where TM : notnull
{
    private readonly Func<TM, Task> _persister;
    private readonly Func<TM, TM> _snapshotter;
    private readonly Dictionary<string, Func<TM, string?>> _validators;
    private readonly Func<TM, IReadOnlyDictionary<string, string?>>? _modelValidator;
    private readonly Func<TM, TM>? _resetOnApproved;
    private readonly bool _strict;
    private readonly IMessageHub _hub;
    private readonly Subject<TM> _onApproved = new();
    private readonly Subject<Exception> _approveErrors = new();
    private readonly object _approveErrorGate = new();
    private readonly Subject<IReadOnlyDictionary<string, string>> _errorsChanged = new();
    private readonly Subject<Unit> _canExecuteChangedTrigger = new();
    private readonly object _stateGate = new();

    private TM _model;
    private TM _snapshot;
    private Dictionary<string, string> _errors;
    private volatile bool _disposed;
    private bool _approvalPublishing;
    private bool _approvalPrePublishing;
    private int _approvalPublisherThreadId;
    private readonly Queue<TM> _deferredApprovalModels = new();
    private int _activeMutations;
    private bool _mutationTeardownPending;

    // ── Builder factory ───────────────────────────────────────────────────────

    /// <summary>
    /// Returns an empty <see cref="FormVMBuilder{TM}"/> for fluent construction.
    /// See ADR-0035 §2 FV1 / FV2.
    /// </summary>
    public static FormVMBuilder<TM> Builder() => FormVMBuilder<TM>.Empty;

    // ── Constructors ──────────────────────────────────────────────────────────

    /// <summary>
    /// Creates a <see cref="FormVM{TM}"/> with a delegate persister.
    /// </summary>
    /// <param name="initial">Initial domain model value; also becomes the initial Snapshot.</param>
    /// <param name="persister">Consumer-supplied async persist delegate. Throw on failure.</param>
    /// <param name="hub">Message hub for DenyCommand hub messages. Defaults to <see cref="NullMessageHub.Instance"/>.</param>
    /// <param name="strict">
    /// When <c>true</c>, <see cref="ApproveCommand"/>.<c>CanExecute</c> returns <c>false</c> when
    /// <see cref="IsDirty"/> is <c>false</c>. Defaults to <c>false</c>.
    /// </param>
    /// <param name="snapshotter">
    /// Custom snapshot function. Defaults to a deep clone via a
    /// <see cref="System.Text.Json"/> round-trip (so a nested-object mutation is
    /// tracked by <see cref="IsDirty"/> and reverted by <see cref="DenyCommand"/>).
    /// Inject this as the escape hatch for models JSON cannot round-trip.
    /// </param>
    /// <param name="validators">Optional field validators keyed by field/property name.</param>
    /// <param name="modelValidator">Optional model-level validator returning field-name errors.</param>
    /// <param name="resetOnApproved">Optional post-persist callback that derives the next pristine model.</param>
    public FormVM(
        TM initial,
        Func<TM, Task> persister,
        IMessageHub? hub = null,
        bool strict = false,
        Func<TM, TM>? snapshotter = null,
        IReadOnlyDictionary<string, Func<TM, string?>>? validators = null,
        Func<TM, IReadOnlyDictionary<string, string?>>? modelValidator = null,
        Func<TM, TM>? resetOnApproved = null)
    {
        ThrowHelper.ThrowIfNull(initial, nameof(initial));
        ThrowHelper.ThrowIfNull(persister, nameof(persister));

        _persister = persister;
        _hub = hub ?? NullMessageHub.Instance;
        _strict = strict;
        _snapshotter = snapshotter ?? DefaultSnapshotter;
        _validators = CopyValidators(validators);
        _modelValidator = modelValidator;
        _resetOnApproved = resetOnApproved;

        _model = initial;
        _snapshot = _snapshotter(initial);
        _errors = Validate(initial);

        IObservable<Unit> canExecuteTrigger = _canExecuteChangedTrigger;

        DenyCommand = RelayCommand.Builder()
            .Task(Deny)
            .Build();

        ApproveCommand = RelayCommand.Builder()
            .Task(ApproveCommandExecute)
            .Predicate(() => IsValid && (!_strict || IsDirty))
            .Triggers(canExecuteTrigger)
            .Build();
    }

    // Fire-and-forget command entry-point. ApproveCommand.Execute() is void, so a
    // persister failure cannot propagate to the caller; route the faulted Task to
    // the ApproveErrors channel instead of discarding it (a discarded faulted Task
    // would otherwise surface only as an UnobservedTaskException at GC) (VMX-008).
    // Await ApproveAsync() to observe the failure inline.
    private void ApproveCommandExecute()
    {
        _ = ApproveInternalAsync().ContinueWith(
            t =>
            {
                EmitApproveError(t.Exception!.GetBaseException());
            },
            CancellationToken.None,
            TaskContinuationOptions.OnlyOnFaulted | TaskContinuationOptions.ExecuteSynchronously,
            TaskScheduler.Default);
    }

    /// <summary>
    /// Creates a <see cref="FormVM{TM}"/> with an <see cref="IFormPersister{TM}"/> interface collaborator.
    /// </summary>
    public FormVM(
        TM initial,
        IFormPersister<TM> persister,
        IMessageHub? hub = null,
        bool strict = false,
        Func<TM, TM>? snapshotter = null,
        IReadOnlyDictionary<string, Func<TM, string?>>? validators = null,
        Func<TM, IReadOnlyDictionary<string, string?>>? modelValidator = null,
        Func<TM, TM>? resetOnApproved = null)
        : this(
            initial,
            persister is null ? throw new ArgumentNullException(nameof(persister)) : (Func<TM, Task>)(model => persister.PersistAsync(model)),
            hub,
            strict,
            snapshotter,
            validators,
            modelValidator,
            resetOnApproved)
    {
    }

    // ── Properties ────────────────────────────────────────────────────────────

    /// <summary>Live, editable model. Set via <see cref="SetModel"/> to mutate.</summary>
    public TM Model { get { lock (_stateGate) return _model; } }

    /// <summary>Read-only snapshot captured at construction (updated after a successful Approve).</summary>
    public TM Snapshot { get { lock (_stateGate) return _snapshot; } }

    /// <summary>
    /// <c>true</c> when <see cref="Model"/> is structurally not equal to <see cref="Snapshot"/>.
    /// Uses <see cref="object.Equals(object?, object?)"/>; record types provide structural equality.
    /// </summary>
    public bool IsDirty { get { lock (_stateGate) return !Equals(_model, _snapshot); } }

    /// <summary>Current validation errors keyed by field/property name.</summary>
    public IReadOnlyDictionary<string, string> Errors
    {
        get { lock (_stateGate) return new Dictionary<string, string>(_errors); }
    }

    /// <summary><c>true</c> when the current model has no validation errors.</summary>
    public bool IsValid { get { lock (_stateGate) return _errors.Count == 0; } }

    /// <summary>
    /// Reverts <see cref="Model"/> to <see cref="Snapshot"/> and publishes
    /// <see cref="FormRevertedMessage"/> + <see cref="PropertyChangedMessage{TSender}"/> on the hub.
    /// </summary>
    public ICommand DenyCommand { get; }

    /// <summary>
    /// Invokes the persister delegate; on success advances <see cref="Snapshot"/> and fires <see cref="OnApproved"/>.
    /// On failure, no state mutation occurs; the persister exception is observable
    /// via <see cref="ApproveAsync"/> only — <c>ApproveCommand.Execute()</c>
    /// dispatches the persist call fire-and-forget (per FORM-007 and chapter 20 §2).
    /// In strict mode, <c>CanExecute</c> is <c>false</c> when <see cref="IsDirty"/> is <c>false</c>.
    /// </summary>
    public ICommand ApproveCommand { get; }

    /// <summary>
    /// Observable that emits the persisted model value after each successful persist.
    /// </summary>
    public IObservable<TM> OnApproved => _onApproved.AsObservable();

    /// <summary>
    /// Observable that surfaces the persister exception when the approve
    /// <em>command</em> (<c>ApproveCommand.Execute()</c>) fails. That path is
    /// fire-and-forget (<see cref="ICommand.Execute"/> is <c>void</c>), so the
    /// failure cannot propagate to the caller; it is emitted here instead of
    /// being discarded with the faulted <see cref="Task"/> (VMX-008). The
    /// awaitable <see cref="ApproveAsync"/> path keeps its throw behavior — await
    /// it directly to handle the error inline. Completes on <see cref="Dispose"/>.
    /// </summary>
    public IObservable<Exception> ApproveErrors => _approveErrors.AsObservable();

    /// <summary>Observable that emits when the effective validation error map changes.</summary>
    public IObservable<IReadOnlyDictionary<string, string>> ErrorsChanged => _errorsChanged.AsObservable();

    /// <summary>Returns the current validation error for a field, if any.</summary>
    public string? FieldError(string field)
    {
        lock (_stateGate)
            return _errors.TryGetValue(field, out var error) ? error : null;
    }

    /// <summary>
    /// Awaitable entry-point to the approve flow. Invokes the persister, advances
    /// <see cref="Snapshot"/> on success, and fires <see cref="OnApproved"/>.
    /// Throws when the persister throws (no state mutation occurs on failure).
    /// Use this in tests and orchestrators that need to await the full cycle.
    /// </summary>
    public Task ApproveAsync() => ApproveInternalAsync();

    // ── Public mutation ───────────────────────────────────────────────────────

    /// <summary>
    /// Replaces the current <see cref="Model"/>. Updating the model may change <see cref="IsDirty"/>;
    /// in strict mode it also fires <see cref="ICommand.CanExecuteChanged"/> on <see cref="ApproveCommand"/>.
    /// </summary>
    public void SetModel(TM model)
    {
        // Inert after Dispose (like ApproveAsync/Deny): a post-dispose call would
        // otherwise emit on disposed validation subjects and throw
        // ObjectDisposedException (parity with the TS/Swift no-op).
        var caller = Environment.CurrentManagedThreadId;
        var admitted = false;
        try
        {
            lock (_stateGate)
            {
                while (_approvalPublishing && _approvalPublisherThreadId != caller && !_disposed)
                    Monitor.Wait(_stateGate);
                if (_disposed) return;
                if (_approvalPublishing && _approvalPrePublishing)
                {
                    _deferredApprovalModels.Enqueue(model);
                    return;
                }
                _activeMutations++;
                admitted = true;
                ThrowHelper.ThrowIfNull(model, nameof(model));
                if (Equals(_model, model)) return;
                var wasDirty = !Equals(_model, _snapshot);
                var wasValid = _errors.Count == 0;
                _model = model;
                var nextErrors = Validate(model);
                var isDirty = !Equals(model, _snapshot);
                var errorsChanged = !SameErrors(nextErrors, _errors);
                _errors = nextErrors;
                var canExecuteChanged = (_strict && isDirty != wasDirty) ||
                    (_errors.Count == 0) != wasValid;
                if (errorsChanged)
                    _errorsChanged.OnNext(new Dictionary<string, string>(nextErrors));
                if (canExecuteChanged)
                    _canExecuteChangedTrigger.OnNext(Unit.Default);
            }
            _hub.Send(PropertyChangedMessage<FormVM<TM>>.Create(
                this,
                nameof(FormVM<TM>),
                nameof(Model)));
        }
        finally
        {
            if (admitted) EndMutation();
        }
    }

    // ── IDisposable ───────────────────────────────────────────────────────────

    /// <summary>Completes the <see cref="OnApproved"/> observable and disposes resources.</summary>
    public void Dispose()
    {
        var caller = Environment.CurrentManagedThreadId;
        var tearDown = false;
        lock (_stateGate)
        {
            while (_approvalPublishing && _approvalPublisherThreadId != caller && !_disposed)
                Monitor.Wait(_stateGate);
            if (_disposed) return;
            _disposed = true;
            if (_activeMutations == 0)
                tearDown = true;
            else
                _mutationTeardownPending = true;
        }
        if (tearDown) TearDown();
    }

    private void EndMutation()
    {
        var tearDown = false;
        lock (_stateGate)
        {
            _activeMutations--;
            if (_activeMutations == 0 && _mutationTeardownPending)
            {
                _mutationTeardownPending = false;
                tearDown = true;
            }
        }
        if (tearDown) TearDown();
    }

    private void TearDown()
    {
        lock (_approveErrorGate)
        {
            _approveErrors.OnCompleted();
            _approveErrors.Dispose();
        }
        _onApproved.OnCompleted();
        _onApproved.Dispose();
        _errorsChanged.OnCompleted();
        _errorsChanged.Dispose();
        _canExecuteChangedTrigger.OnCompleted();
        _canExecuteChangedTrigger.Dispose();
        if (DenyCommand is IDisposable d1) d1.Dispose();
        if (ApproveCommand is IDisposable d2) d2.Dispose();
    }

    private void EmitApproveError(Exception error)
    {
        lock (_approveErrorGate)
        {
            if (_disposed) return;
            _approveErrors.OnNext(error);
        }
    }

    // ── Internal ──────────────────────────────────────────────────────────────

    private void Deny()
    {
        var caller = Environment.CurrentManagedThreadId;
        var admitted = false;
        try
        {
            bool canExecuteChanged;
            lock (_stateGate)
            {
                while (_approvalPublishing && _approvalPublisherThreadId != caller && !_disposed)
                    Monitor.Wait(_stateGate);
                if (_disposed) return;
                _activeMutations++;
                admitted = true;
                var wasDirty = !Equals(_model, _snapshot);
                var wasValid = _errors.Count == 0;
                var nextModel = _snapshotter(_snapshot);
                _model = nextModel;
                var nextErrors = Validate(nextModel);
                var isDirty = !Equals(nextModel, _snapshot);
                var errorsChanged = !SameErrors(nextErrors, _errors);
                _errors = nextErrors;
                var isValid = _errors.Count == 0;

                if (errorsChanged)
                    _errorsChanged.OnNext(new Dictionary<string, string>(nextErrors));
                canExecuteChanged = (_strict && wasDirty != isDirty) || isValid != wasValid;
            }

            _hub.Send(new FormRevertedMessage(this, nameof(FormVM<TM>)));
            _hub.Send(PropertyChangedMessage<FormVM<TM>>.Create(this, nameof(FormVM<TM>), nameof(Model)));
            if (canExecuteChanged)
            {
                lock (_stateGate)
                    _canExecuteChangedTrigger.OnNext(Unit.Default);
            }
        }
        finally
        {
            if (admitted) EndMutation();
        }
    }

    private async Task ApproveInternalAsync()
    {
        // A disposed form is a full no-op — the persister must not be
        // invoked (symmetric with the Deny guard).
        TM current;
        lock (_stateGate)
        {
            if (_disposed || _errors.Count != 0) return;
            current = _model;
        }

        // May throw — intentional. No state mutation if this throws.
        await _persister(current).ConfigureAwait(false);

        // Dispose() may have run during the await; the subjects below are
        // completed and disposed, so emitting would throw inside an
        // unobserved task.
        lock (_stateGate)
        {
            if (_disposed) return;
        }

        // Success: either atomically install the configured pristine reset
        // state, or preserve the legacy snapshot-advance behavior.
        TM? nextModel = default;
        TM nextSnapshot;
        Dictionary<string, string>? nextErrors = null;
        if (_resetOnApproved is not null)
        {
            // Prepare everything before assigning any field. A reset or
            // snapshotter failure therefore leaves local state untouched even
            // though persistence has already succeeded.
            var reset = _resetOnApproved(current);
            nextModel = _snapshotter(reset);
            nextSnapshot = _snapshotter(reset);
            nextErrors = Validate(nextModel);
        }
        else
        {
            nextSnapshot = _snapshotter(current);
        }

        var caller = Environment.CurrentManagedThreadId;
        bool errorsChanged;
        bool canExecuteChanged;
        IReadOnlyDictionary<string, string>? publishedErrors = null;
        lock (_stateGate)
        {
            while (_approvalPublishing && _approvalPublisherThreadId != caller)
                Monitor.Wait(_stateGate);
            if (_disposed) return;
            _approvalPublishing = true;
            _approvalPrePublishing = true;
            _approvalPublisherThreadId = caller;
            var wasDirty = !Equals(_model, _snapshot);
            var wasValid = _errors.Count == 0;
            errorsChanged = false;
            if (_resetOnApproved is not null)
            {
                _model = nextModel!;
                _snapshot = nextSnapshot;
                if (!SameErrors(nextErrors!, _errors))
                {
                    _errors = nextErrors!;
                    errorsChanged = true;
                    publishedErrors = new Dictionary<string, string>(_errors);
                }
            }
            else
            {
                _snapshot = nextSnapshot;
            }
            canExecuteChanged = (_strict && !Equals(_model, _snapshot) != wasDirty) ||
                (_errors.Count == 0) != wasValid;
        }

        try
        {
            if (errorsChanged)
            {
                _errorsChanged.OnNext(publishedErrors!);
                if (_disposed) return;
            }
            if (canExecuteChanged)
            {
                _canExecuteChangedTrigger.OnNext(Unit.Default);
                if (_disposed) return;
            }
            lock (_stateGate)
                _approvalPrePublishing = false;
            _onApproved.OnNext(current);
        }
        finally
        {
            TM[] deferredModels;
            lock (_stateGate)
            {
                _approvalPrePublishing = false;
                _approvalPublishing = false;
                _approvalPublisherThreadId = 0;
                deferredModels = [.. _deferredApprovalModels];
                _deferredApprovalModels.Clear();
                Monitor.PulseAll(_stateGate);
            }
            foreach (var deferredModel in deferredModels)
                SetModel(deferredModel);
        }
    }

    // Cached per closed generic (one instance per TM): System.Text.Json caches
    // serialization metadata against the options instance, so reusing a single
    // instance avoids a per-call metadata rebuild (and CA1869).
    private static readonly System.Text.Json.JsonSerializerOptions DefaultSnapshotJsonOptions =
        new() { IncludeFields = true };

    /// <summary>
    /// Default deep-clone snapshotter: a <see cref="System.Text.Json"/> serialize→deserialize
    /// round-trip. A deep clone is required so the snapshot does not share nested
    /// mutable state with the live model — otherwise an in-place mutation of a
    /// nested object would be invisible to <see cref="IsDirty"/> and could not be
    /// reverted by <see cref="DenyCommand"/> (VMX-064). <c>System.Text.Json</c> is
    /// in the BCL, so no extra package is taken.
    /// <para>
    /// For models JSON cannot round-trip (delegates, cyclic graphs, non-public or
    /// non-default-constructible members, live handles, …) inject a custom
    /// <c>snapshotter</c> via the constructor or builder — that hook is the
    /// documented escape hatch and always overrides this default.
    /// </para>
    /// </summary>
    private static TM DefaultSnapshotter(TM model)
    {
        var json = System.Text.Json.JsonSerializer.Serialize(model, DefaultSnapshotJsonOptions);
        return System.Text.Json.JsonSerializer.Deserialize<TM>(json, DefaultSnapshotJsonOptions)!;
    }

    private Dictionary<string, string> Validate(TM model)
    {
        var errors = new Dictionary<string, string>();
        foreach (var kvp in _validators)
        {
            var error = kvp.Value(model);
            if (error is not null)
                errors[kvp.Key] = error;
        }

        if (_modelValidator is not null)
        {
            foreach (var kvp in _modelValidator(model))
            {
                if (kvp.Value is null)
                    errors.Remove(kvp.Key);
                else
                    errors[kvp.Key] = kvp.Value;
            }
        }

        return errors;
    }

    private static bool SameErrors(
        Dictionary<string, string> left,
        Dictionary<string, string> right)
    {
        if (left.Count != right.Count) return false;
        foreach (var kvp in left)
        {
            if (!right.TryGetValue(kvp.Key, out var other) || other != kvp.Value)
                return false;
        }

        return true;
    }

    private static Dictionary<string, Func<TM, string?>> CopyValidators(
        IReadOnlyDictionary<string, Func<TM, string?>>? validators)
    {
        var copy = new Dictionary<string, Func<TM, string?>>();
        if (validators is null) return copy;
        foreach (var kvp in validators)
            copy[kvp.Key] = kvp.Value;
        return copy;
    }
}
